### ISSUE TITLE
[Proposal]: Unity receipt format

### ISSUE THREAD

**Issue by Mazyod** [voltrue2/in-app-purchase#69]
Hello @voltrue2,

Thanks for making this great project! It really saved me a lot of time in integrating receipt verification for my Unity game.

I wanted to propose a feature, which I can also implement and send a PR if you agree.

Unity has a unified IAP module which returns a specific format, which [you can read about here][unity-docs]. It is something like this:

```js
{"Store":"GooglePlay","TransactionID":"...","Payload":"{\"json\":\"{\\\"packageName\\\":\\\"...\\\",\\\"productId\\\":\\\"...\\\",\\\"purchaseTime\\\":1479474689471,\\\"purchaseState\\\":0,\\\"purchaseToken\\\":\\\"...\\\"}\",\"signature\":\"...\"}"}
```

As you can see, the data already contains the store type and the receipt payload from the store. However, I don't know why, they have `"json"` as a key instead of `"data"`, which makes the verification fail.

So, will it be useful if I submit a PR, something like `iap.validate(iap.UNITY, unity_data, ...)`? This function will automatically detect the store, and also correct the `"json"` key to `"data"` for Google Play.

Of course, if you think this is out of scope of this library, I totally understand and will implement this feature on my server code instead.

[unity-docs]: https://docs.unity3d.com/Manual/UnityIAPPurchaseReceipts.html

**Comment by voltrue2** [comment]
Hello @Mazyod 

Thank you for reaching out. The idea is certainly an interesting one and I understand it is rather annoying to have to manually check and change the key from json to data.

Where is a question for you though, when you use iap.UNITY, would it support iOS also?

**Comment by Mazyod** [comment]
@voltrue2 of course! In fact, I will test and make sure it will support Apple, Google, Amazon, & MS.

**Comment by Mazyod** [comment]
@voltrue2 I spent some time running the tests and attempting to integrate `iap.UNITY` verification strategy, but it has become clear to me that the current library design is a bit rigid and doesn't allow us to easily implement a layer of indirection very easily.

I mean, consider this. I have multiple games in Unity, and I would like to dynamically choose which one it is. With the current design, this means, I have to send `iap.UNITY` the dynamic configuration for **all** stores.

Also, by looking at the tests and how this library evolved, it might need some internal refactoring (without major API breaking changes) to make it easier to improve. I think this takes priority now over implementing `UNITY` strategy. I will try to help out there, but I can't make any promises for now.

Does that sound reasonable to you?

**Comment by voltrue2** [comment]
Hello

Thank you for taking the time on this. The initial idea of this library was to provide a very simple and uniformed validation API for all of these platforms. It has changed over time and along with the code. 

Anyways yes I agree with your proposal. I've been wanting to refactor the code for sometimes anyways.

**Comment by axhar** [comment]
Hey Guys! is there any update on this? @Mazyod did spend time on it to make like you proposed.

I came across this library in a need to implement something that matches you use case of using a single server for multiple games and on multiple platforms. It will be great help if you can share some insights from your experience.

**Comment by Mazyod** [comment]
I have two PRs open, but I am not sure if @voltrue2 is interested in pursuing this idea since it introduces relatively large structural changes. I just went with the hack I mentioned above for my service, and this library works great.

**Comment by voltrue2** [comment]
Hello,

I apologize for the long silence.
I will be looking at the PRs and follow up soon.

Cheers!

**Comment by voltrue2** [comment]
Hello,

Finally I am working on adding the support for Unity receipts.

Does anyone have a valid unity receipt available for me to test?

Cheers.

**Comment by Mazyod** [comment]
Here are the tests you need to add:

```elixir
%{"Store" => _, "TransactionID" => _, "Payload" => _}
```

Where `Store` I map it as follows:

```elixir
%{"MacAppStore" => "apple", "AppleAppStore" => "apple", "GooglePlay" => "google"}
```

And as I mentioned, the `Payload` is the receipt I send to this library, except if it's google, I process it first by replacing `{"json"` at the beginning, to `{"data"`, where Unity returns `json`, and this library expects `data`.

**Comment by voltrue2** [comment]
Thanks. I just realized that I can create Unity format receipt myself lol.
My implementation already takes care of preprocessing of Payload for Google.
I will create tests and hopefully I could push this update finally.

Thank you again for the suggestion and pointers!

Sorry to keep you waiting on this for so son.

Cheers!

**Comment by voltrue2** [comment]
Hello
I have added tests and fixed bugs.
Here is the commit: 9b1a35e7d3b39f2363680f7e32af34db4c4d035c
I will be publishing this version as `1.7.0` on npm.

Thanks again for the help.

Cheers

**Comment by Mazyod** [comment]
I don't get this:
`if (receipt.Store === constants.UNITY.GOOGLE || constants.UNITY.APPLE)`

I think you must check against apple receipt, instead of just checking if the apple constant is not null or undefined:
`receipt.Store === constants.UNITY.APPLE`

**Comment by voltrue2** [comment]
@Mazyod 
If you check against `constants.UNITY.APPLE` only, it will not work for Google Play.

**Comment by Mazyod** [comment]
I mean:
```js
// from this
if (receipt.Store === constants.UNITY.GOOGLE || constants.UNITY.APPLE)
// to this
if (receipt.Store === constants.UNITY.GOOGLE || receipt.Store === constants.UNITY.APPLE)
```

**Comment by voltrue2** [comment]
fix: a1317bd388d8836e18b3ed6353af24b5e6f4b796

**Comment by voltrue2** [comment]
@Mazyod Thanks for the help.

I have now published the newer version with the fix on npm.

Cheers!

**Comment by Skjalgsm** [comment]
Am I correct in my assessment that automatic detection of receipts does not work when passing in a unity receipt? I have to manually specify that it is a unity receipt?

**Comment by voltrue2** [comment]
No, the module should auto-detect a unity receipt also.
If that is not a case, it’s a bug. Would you mind sharing the details?

Cheers

**Comment by Skjalgsm** [comment]
When I try to use validate or validateOnce with a unity receipt with a GooglePlay receipt inside it it tries to validate it with amazon.

The problem is line 168
`if (service === module.exports.UNITY) {`

This is the first thing that happens inside validateOnce when passing in a unity receipt and at this stage the service parameter is undefined so this check fails and thus it does not detect the correct service or the correct receipt.

**Comment by voltrue2** [comment]
@Skjalgsm 
Sorry for getting back to you late on this issue here.
I have tested `.validateOnce` with public key, but I cannot reproduce the issue you are talking about.
Would you mind sharing more details with code example?

Cheers
